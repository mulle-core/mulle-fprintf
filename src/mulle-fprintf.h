/*
 * template: demo/library/PROJECT_SOURCE_DIR/PROJECT_NAME.h
 * vendor/extension: mulle-sde/c-demo
 */
#ifndef mulle_fprintf_h__
#define mulle_fprintf_h__

#include "include.h"

#include <stdint.h>

/*
 *  (c) 2021 Nat! Mulle kybernetiK 
 *
 *  version:  major, minor, patch
 */
#define MULLE_FPRINTF_VERSION  ((0 << 20) | (0 << 8) | 1)


static inline unsigned int   mulle_fprintf_get_version_major( void)
{
   return( MULLE_FPRINTF_VERSION >> 20);
}


static inline unsigned int   mulle_fprintf_get_version_minor( void)
{
   return( (MULLE_FPRINTF_VERSION >> 8) & 0xFFF);
}


static inline unsigned int   mulle_fprintf_get_version_patch( void)
{
   return( MULLE_FPRINTF_VERSION & 0xFF);
}

MULLE_FPRINTF_EXTERN_GLOBAL
uint32_t   mulle_fprintf_get_version( void);


#include <stdio.h>
#include <stdarg.h>

MULLE_FPRINTF_EXTERN_GLOBAL
int   mulle_printf( char *format, ...);

MULLE_FPRINTF_EXTERN_GLOBAL
int   mulle_vprintf( char *format, va_list args);

MULLE_FPRINTF_EXTERN_GLOBAL
int   mulle_fprintf( FILE *fp, char *format, ...);

MULLE_FPRINTF_EXTERN_GLOBAL
int   mulle_vfprintf( FILE *fp, char *format, va_list args);


/*
 * The following files are auto-generated.
 * The _mulle-fprintf-provide header is generated by
 * mulle-match-to-c during a reflect.
 */
#include "_mulle-fprintf-provide.h"

/*
 * The versioncheck header can be generated with
 * mulle-project-dependency-versions, but it is optional.
 */
#ifdef __has_include
# if __has_include( "_mulle-fprintf-versioncheck.h")
#  include "_mulle-fprintf-versioncheck.h"
# endif
#endif

#endif
