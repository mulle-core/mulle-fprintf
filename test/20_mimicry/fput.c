#include <mulle-fprintf/mulle-fprintf.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <setjmp.h>

static jmp_buf jump_buffer;

static void signal_handler(int signum)
{
   // MEMO: this is wrong, you can't really do this (but works in this test for now)
    longjmp(jump_buffer, signum);
}

static int safe_fputc(int c, FILE *fp)
{
    int result;
    if (setjmp(jump_buffer) == 0)
    {
        result = fputc(c, fp);
    }
    else
    {
        result = -2;  // Special value to indicate crash
    }
    return result;
}

static int safe_mulle_buffer_fputc(int c, void *buffer)
{
    int result;
    if (setjmp(jump_buffer) == 0)
    {
        result = mulle_buffer_fputc(c, buffer);
    }
    else
    {
        result = -2;  // Special value to indicate crash
    }
    return result;
}

static void compare_putc(int c, FILE *fp, void *buffer, const char *test_name)
{
    int r1, r2;
    int errno1, errno2;

    signal(SIGSEGV, signal_handler);

    errno = 0;
    r1 = safe_fputc(c, fp);
    errno1 = errno;

    errno = 0;
    r2 = safe_mulle_buffer_fputc(c, buffer);
    errno2 = errno;

    if (r1 == r2 && errno1 == errno2)
    {
        if (r1 == -2)
        {
            printf("%s: Passed (both crashed as expected)\n", test_name);
        }
        else
        {
            printf("%s: Passed (result=%d, errno=%d)\n", test_name, r1, errno1);
        }
    }
    else
    {
        printf("Error in %s: fputc: result=%d, errno=%d; mulle_buffer_fputc: result=%d, errno=%d\n",
               test_name, r1, errno1, r2, errno2);
    }

    signal(SIGSEGV, SIG_DFL);  // Reset signal handler
}

static void test_putc(void)
{
    FILE *fp;
    void *buffer;
    char test_char = 'A';

    // Test case 1: NULL pointers
    compare_putc(test_char, NULL, NULL, "NULL pointers");

    // Test case 2: Write to memory stream
    fp = fmemopen(NULL, 100, "w");
    buffer = mulle_buffer_fmemopen(NULL, 100, "w");
    compare_putc(test_char, fp, buffer, "Write to memory stream");
    fclose(fp);
    mulle_buffer_fclose(buffer);

    // Test case 3: Write to full buffer
    char small_buf[1] = {0};
    fp = fmemopen(small_buf, 1, "w");
    buffer = mulle_buffer_fmemopen(small_buf, 1, "w");
    compare_putc(test_char, fp, buffer, "Write to full buffer (first write)");
    compare_putc(test_char, fp, buffer, "Write to full buffer (second write)");
    fclose(fp);
    mulle_buffer_fclose(buffer);

    // Test case 4: Write-only empty file
    fp = fmemopen(NULL, 0, "w");
    buffer = mulle_buffer_fmemopen(NULL, 0, "w");
    compare_putc(test_char, fp, buffer, "Write-only empty file");
    fclose(fp);
    mulle_buffer_fclose(buffer);

    // Test case 5: Multiple writes
    fp = fmemopen(NULL, 100, "w");
    buffer = mulle_buffer_fmemopen(NULL, 100, "w");
    for (int i = 0; i < 5; i++)
    {
        compare_putc('0' + i, fp, buffer, "Multiple writes");
    }
    fclose(fp);
    mulle_buffer_fclose(buffer);
}

int main(int argc, const char * argv[])
{
    test_putc();
    return 0;
}
