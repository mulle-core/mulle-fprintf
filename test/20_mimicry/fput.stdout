Error in NULL pointers: fputc: result=-2, errno=0; mulle_buffer_fputc: result=-1, errno=0
Write to memory stream: Passed (result=65, errno=0)
Write to full buffer (first write): Passed (result=65, errno=0)
Write to full buffer (second write): Passed (result=65, errno=0)
Write-only empty file: Passed (result=65, errno=0)
Multiple writes: Passed (result=48, errno=0)
Multiple writes: Passed (result=49, errno=0)
Multiple writes: Passed (result=50, errno=0)
Multiple writes: Passed (result=51, errno=0)
Multiple writes: Passed (result=52, errno=0)
